File: src/nanodoc/v2/renderer.py
"""Rendering for Nanodoc v2.

This module handles the "Rendering" stage of the Nanodoc v2 pipeline.
It takes a Document object and generates the final output string,
including basic concatenation of FileContent content and TOC generation.
"""

import os
import re

from nanodoc.v2.formatter import (
    enhance_rendering,
    format_with_line_numbers,
)
from nanodoc.v2.structures import Document


def render_document(
    document: Document, include_toc: bool = False, include_line_numbers: bool = False
) -> str:
    """Render a Document object to a string.

    This function:
    - Generates a table of contents if requested
    - Concatenates the content of all FileContent objects
    - Adds file separators between non-inlined content
    - Optionally adds line numbers

    Args:
        document: Document object to render
        include_toc: Whether to include a table of contents
        include_line_numbers: Whether to include line numbers

    Returns:
        Rendered document as a string
    """
    rendered_parts = []

    # Generate TOC if requested
    if include_toc:
        toc = generate_toc(document)
        if toc:
            rendered_parts.append(toc)

    # Concatenate content
    prev_original_source = None
    for item in document.content_items:
        # Add file separator if needed
        # Check if not inlined and different from previous
        is_not_inlined = not item.original_source
        different_source = item.filepath != prev_original_source

        if is_not_inlined and different_source:
            # Add a separator if this isn't the first content item
            if rendered_parts and not rendered_parts[-1].endswith("\n\n"):
                rendered_parts.append("\n")

            # Add file header
            file_basename = os.path.basename(item.filepath)
            rendered_parts.append(f"# {file_basename}\n\n")

        # Add the content with optional line numbers
        content_to_add = item.content
        if include_line_numbers:
            # Use the formatter's line numbering function
            content_to_add = format_with_line_numbers(content_to_add)

        rendered_parts.append(content_to_add)

        # Ensure content ends with a newline
        if rendered_parts and not rendered_parts[-1].endswith("\n"):
            rendered_parts.append("\n")

        # Track the source for the next iteration
        prev_original_source = item.original_source or item.filepath

    # Join all parts to create the final content
    plain_content = "".join(rendered_parts)

    # Apply theming if requested
    if hasattr(document, "use_rich_formatting") and document.use_rich_formatting:
        return enhance_rendering(
            plain_content,
            theme_name=document.theme_name,
            use_rich_formatting=document.use_rich_formatting,
        )

    return plain_content


def generate_toc(document: Document) -> str:
    """Generate a table of contents from a Document.

    Args:
        document: Document object to generate TOC for

    Returns:
        Table of contents as a string
    """
    # Extract headings from content
    headings = _extract_headings(document)

    if not headings:
        return ""

    toc_lines = ["# Table of Contents\n\n"]

    for file_path, file_headings in headings.items():
        # Add file entry
        filename = os.path.basename(file_path)
        toc_lines.append(f"- {filename}\n")

        # Add headings for this file
        for heading, _ in file_headings:
            # Indent heading
            toc_lines.append(f"  - {heading}\n")

    toc_lines.append("\n")  # Add blank line after TOC

    # Store TOC data in the document for future reference
    document.toc = headings

    return "".join(toc_lines)


def _extract_headings(document: Document) -> dict[str, list[tuple[str, int]]]:
    """Extract headings from document content.

    Args:
        document: Document object to extract headings from

    Returns:
        Dictionary mapping file paths to lists of (heading_text,
        line_number) tuples
    """
    headings_by_file = {}

    # Markdown heading regex (# Heading)
    heading_pattern = re.compile(r"^(#+)\s+(.+)$", re.MULTILINE)

    for item in document.content_items:
        file_headings = []

        # Use the original source if available, otherwise use the filepath
        file_path = item.original_source or item.filepath

        # Extract headings with line numbers
        lines = item.content.split("\n")
        for i, line in enumerate(lines):
            match = heading_pattern.match(line)
            if match:
                heading_level = len(match.group(1))  # Number of # characters
                heading_text = match.group(2).strip()

                # Only include level 1 and 2 headings
                if heading_level <= 2:
                    file_headings.append((heading_text, i + 1))

        # Store headings for this file if any were found
        if file_headings:
            if file_path in headings_by_file:
                headings_by_file[file_path].extend(file_headings)
            else:
                headings_by_file[file_path] = file_headings

    return headings_by_file


# This function is no longer used directly - formatter.format_with_line_numbers
# is used instead, but keeping it here to avoid breaking tests
def _add_line_numbers(content: str) -> str:
    """Add line numbers to content.

    Args:
        content: Content to add line numbers to

    Returns:
        Content with line numbers
    """
    return format_with_line_numbers(content)

File: pkg/nanodoc/renderer.go
package nanodoc

import (
	"fmt"
	"log/slog"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/arthur-debert/nanodoc/pkg/markdown"
)

// RendererOptions controls how the document is rendered
type RendererOptions struct {
	IncludeTOC         bool
	IncludeLineNumbers bool
	UseRichFormatting  bool
}

// RenderDocument renders a Document object to a string
func RenderDocument(doc *Document, ctx *FormattingContext) (string, error) {
	// For markdown output, use enhanced renderer with all features
	if doc.FormattingOptions.OutputFormat == "markdown" {
		return renderMarkdownEnhanced(doc, ctx)
	}

	// For plain output, concatenate without any formatting
	if doc.FormattingOptions.OutputFormat == "plain" {
		return renderPlainText(doc)
	}

	var parts []string

	// Generate TOC first, as it's used for filenames
	if ctx.ShowTOC || ctx.HeaderFormat == HeaderFormatNice {
		slog.Debug("Generating table of contents for filenames/TOC")
		generateTOC(doc)
	}

	// Render TOC if requested
	if ctx.ShowTOC {
		var tocParts []string
		tocParts = append(tocParts, "Table of Contents")
		tocParts = append(tocParts, "=================")
		tocParts = append(tocParts, "")
		for _, entry := range doc.TOC {
			// Indent based on heading level, assuming Level 1 is the base
			indent := strings.Repeat("  ", entry.Level-1)
			tocParts = append(tocParts, fmt.Sprintf("%s- %s (%s)", indent, entry.Title, filepath.Base(entry.Path)))
		}
		tocParts = append(tocParts, "")
		parts = append(parts, strings.Join(tocParts, "\n"))
		parts = append(parts, "\n")
	}

	// Render each content item
	prevOriginalSource := ""
	sequenceNumber := 0
	globalLineNumber := 1

	for _, item := range doc.ContentItems {
		// Check if we need a file separator
		isNotInlined := item.OriginalSource == ""
		differentSource := item.Filepath != prevOriginalSource

		if isNotInlined && differentSource && ctx.ShowFilenames {
			// Add separator if not first item
			if len(parts) > 0 && !strings.HasSuffix(parts[len(parts)-1], "\n\n") {
				parts = append(parts, "\n")
			}

			// Generate filename
			sequenceNumber++
			filename := generateFilename(item.Filepath, &doc.FormattingOptions, sequenceNumber, doc)
			parts = append(parts, filename)
			parts = append(parts, "\n\n")
		}

		// Add content with optional line numbers
		content := item.Content
		
		// Handle empty files
		if content == "" {
			content = "(empty file)"
		}
		
		if ctx.LineNumbers != LineNumberNone {
			numberedContent, newGlobalLineNum := addLineNumbers(content, ctx.LineNumbers, globalLineNumber)
			content = numberedContent
			if ctx.LineNumbers == LineNumberGlobal {
				globalLineNumber = newGlobalLineNum
			}
		}

		parts = append(parts, content)

		// Ensure content ends with newline
		if len(parts) > 0 && !strings.HasSuffix(parts[len(parts)-1], "\n") {
			parts = append(parts, "\n")
		}

		// Track source for next iteration
		if item.OriginalSource != "" {
			prevOriginalSource = item.OriginalSource
		} else {
			prevOriginalSource = item.Filepath
		}
	}

	result := strings.Join(parts, "")
	return result, nil
}

func generateFilename(filePath string, opts *FormattingOptions, seqNum int, doc *Document) string {
	headerText := generateFileHeaderText(filePath, opts, seqNum, doc)

	// Get banner style from registry
	style, exists := GetBannerStyle(opts.HeaderStyle)
	if !exists {
		// Fallback to none style if not found
		style, _ = GetBannerStyle("none")
	}

	// Apply the banner style
	return style.Apply(headerText, opts)
}

// generateFileHeaderText generates the text content for a file header
func generateFileHeaderText(filePath string, opts *FormattingOptions, seqNum int, doc *Document) string {
	// Find the primary title for this file from the TOC
	var title string
	for _, entry := range doc.TOC {
		if entry.Path == filePath {
			title = entry.Title
			break
		}
	}

	var baseName string
	switch opts.HeaderFormat {
	case HeaderFormatFilename:
		baseName = filepath.Base(filePath)
	case HeaderFormatPath:
		baseName = filePath
	case HeaderFormatNice:
		fallthrough
	default:
		// Use title from TOC if available, otherwise generate from filename
		niceName := title
		if niceName == "" {
			filename := filepath.Base(filePath)
			nameWithoutExt := strings.TrimSuffix(filename, filepath.Ext(filename))
			niceName = strings.ReplaceAll(nameWithoutExt, "_", " ")
			niceName = strings.ReplaceAll(niceName, "-", " ")
			niceName = splitCamelCase(niceName)
			niceName = toTitleCase(niceName)
		}
		baseName = niceName
	}

	// Add sequence number
	seq := generateSequence(seqNum, opts.SequenceStyle)
	if seq != "" {
		return fmt.Sprintf("%s. %s", seq, baseName)
	}
	return baseName
}

// generateSequence generates a sequence number in the specified style
func generateSequence(num int, style SequenceStyle) string {
	switch style {
	case SequenceNumerical:
		return strconv.Itoa(num)
	case SequenceLetter:
		if num <= 26 {
			return string(rune('a' + num - 1))
		}
		// For numbers > 26, use aa, ab, ac, etc.
		return string(rune('a'+((num-1)/26)-1)) + string(rune('a'+((num-1)%26)))
	case SequenceRoman:
		return toRoman(num)
	default:
		return strconv.Itoa(num)
	}
}

// splitCamelCase splits a camelCase string into words
func splitCamelCase(s string) string {
	// Add space before capital letters preceded by lowercase
	re1 := regexp.MustCompile("([a-z])([A-Z])")
	s = re1.ReplaceAllString(s, "$1 $2")
	
	// Handle consecutive uppercase followed by lowercase (e.g., HTMLFile -> HTML File)
	re2 := regexp.MustCompile("([A-Z])([A-Z][a-z])")
	s = re2.ReplaceAllString(s, "$1 $2")
	
	return s
}

// toTitleCase converts a string to title case (first letter of each word uppercase)
func toTitleCase(s string) string {
	words := strings.Fields(s)
	for i, word := range words {
		if len(word) > 0 {
			words[i] = strings.ToUpper(string(word[0])) + strings.ToLower(word[1:])
		}
	}
	return strings.Join(words, " ")
}

// toRoman converts a number to Roman numerals (simplified version)
func toRoman(num int) string {
	values := []int{1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1}
	symbols := []string{"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"}
	
	result := ""
	for i := 0; i < len(values); i++ {
		for num >= values[i] {
			num -= values[i]
			result += symbols[i]
		}
	}
	return strings.ToLower(result)
}

// addLineNumbers adds line numbers to content
func addLineNumbers(content string, mode LineNumberMode, startNum int) (string, int) {
	lines := strings.Split(content, "\n")
	
	// Calculate the width needed for line numbers
	maxLineNum := startNum + len(lines) - 1
	if mode == LineNumberFile {
		maxLineNum = len(lines)
	}
	width := len(strconv.Itoa(maxLineNum))
	
	var result []string
	lineNum := startNum
	if mode == LineNumberFile {
		lineNum = 1
	}
	
	for _, line := range lines {
		// Don't add line numbers to empty lines at the end
		if line == "" && lineNum == len(lines) {
			result = append(result, line)
		} else {
			numberedLine := fmt.Sprintf("%*d | %s", width, lineNum, line)
			result = append(result, numberedLine)
		}
		lineNum++
	}
	
	return strings.Join(result, "\n"), lineNum
}

// generateTOC generates a table of contents for the document using the markdown parser.
func generateTOC(doc *Document) {
	doc.TOC = make([]TOCEntry, 0)
	parser := markdown.NewParser()
	tocGen := markdown.NewTOCGenerator()

	var allHeadings []TOCEntry
	sequenceNum := 1

	for _, item := range doc.ContentItems {
		// Only extract headings from markdown files
		if !strings.HasSuffix(item.Filepath, ".md") && !strings.HasSuffix(item.Filepath, ".markdown") {
			continue
		}

		mdDoc, err := parser.Parse([]byte(item.Content))
		if err != nil {
			slog.Warn("failed to parse markdown for TOC generation", "file", item.Filepath, "error", err)
			continue
		}

		entries := tocGen.ExtractTOC(mdDoc)
		for _, entry := range entries {
			allHeadings = append(allHeadings, TOCEntry{
				Title:    entry.Text,
				Level:    entry.Level,
				Path:     item.Filepath,
				Sequence: generateSequence(sequenceNum, doc.FormattingOptions.SequenceStyle),
				// LineNumber is not available from the new parser, which is acceptable.
			})
			sequenceNum++
		}
	}
	doc.TOC = allHeadings
}



// renderMarkdownBasic performs basic concatenation of markdown files without any modifications
// This is kept for backward compatibility and fallback
func renderMarkdownBasic(doc *Document) (string, error) {
	var parts []string

	for _, item := range doc.ContentItems {
		// Simply append the content as-is
		parts = append(parts, item.Content)
		
		// Ensure content ends with newline
		if len(parts) > 0 && !strings.HasSuffix(parts[len(parts)-1], "\n") {
			parts = append(parts, "\n")
		}
	}

	result := strings.Join(parts, "")
	return result, nil
}

// renderMarkdownEnhanced uses the markdown package to provide rich markdown output
func renderMarkdownEnhanced(doc *Document, ctx *FormattingContext) (string, error) {
	// Phase 2.1: POC - Demonstrate all capabilities
	parser := markdown.NewParser()
	transformer := markdown.NewTransformer()
	renderer := markdown.NewRenderer()
	tocGen := markdown.NewTOCGenerator()
	headerFormatter := markdown.NewHeaderFormatter()

	var processedDocs []*markdown.Document

	// Generate TOC first if needed, so it's available for all renderers
	if ctx.ShowTOC {
		slog.Debug("Generating table of contents for markdown output")
		generateTOC(doc)
	}

	// Process each content item
	for i, item := range doc.ContentItems {
		mdDoc, err := parser.Parse([]byte(item.Content))
		if err != nil {
			return "", fmt.Errorf("failed to parse content for file %s: %w", item.Filepath, err)
		}

		isMarkdown := strings.HasSuffix(item.Filepath, ".md") || strings.HasSuffix(item.Filepath, ".markdown")

		if isMarkdown {
			// Perform markdown-specific transformations

			// Adjust header levels for subsequent documents to maintain hierarchy
			if i > 0 && transformer.HasH1(mdDoc) {
				if err := transformer.AdjustHeaderLevels(mdDoc, 1); err != nil {
					return "", fmt.Errorf("failed to adjust header levels for %s: %w", item.Filepath, err)
				}
			}

			// Insert file headers if requested
			if ctx.ShowFilenames {
				sequenceNum := i + 1
				headerText := generateFileHeaderText(item.Filepath, &doc.FormattingOptions, sequenceNum, doc)

				// Format as a markdown header. H2 is chosen as a sensible default
				// to avoid conflicting with a potential H1 title in the first document.
				const headerLevel = 2
				mdHeaderText := headerFormatter.FormatFileHeader(headerText, "", headerLevel)

				if err := transformer.InsertFileHeader(mdDoc, mdHeaderText, headerLevel); err != nil {
					return "", fmt.Errorf("failed to insert file header for %s: %w", item.Filepath, err)
				}
			}
		}

		processedDocs = append(processedDocs, mdDoc)
	}

	// Build final output
	var output strings.Builder

	// Phase 2.4: Add TOC if requested
	if ctx.ShowTOC && len(doc.TOC) > 0 {
		// Convert nanodoc.TOCEntry to markdown.TOCEntry
		mdTOCEntries := make([]markdown.TOCEntry, len(doc.TOC))
		for i, entry := range doc.TOC {
			mdTOCEntries[i] = markdown.TOCEntry{
				Text:  fmt.Sprintf("%s - %s", filepath.Base(entry.Path), entry.Title),
				Level: entry.Level,
			}
		}
		tocMarkdown := tocGen.GenerateTOCMarkdown(mdTOCEntries)
		output.WriteString(tocMarkdown)
		output.WriteString("\n\n")
	}

	// Render all processed documents
	for i, mdDoc := range processedDocs {
		if i > 0 {
			output.WriteString("\n")
		}

		rendered, err := renderer.Render(mdDoc)
		if err != nil {
			return "", fmt.Errorf("failed to render markdown: %w", err)
		}

		output.Write(rendered)
	}

	return output.String(), nil
}

// renderPlainText performs basic concatenation without any formatting
func renderPlainText(doc *Document) (string, error) {
	var parts []string

	for _, item := range doc.ContentItems {
		// Simply append the content as-is
		parts = append(parts, item.Content)
		
		// Ensure content ends with newline
		if len(parts) > 0 && !strings.HasSuffix(parts[len(parts)-1], "\n") {
			parts = append(parts, "\n")
		}
	}

	result := strings.Join(parts, "")
	return result, nil
}